File: travel_checklist.go
// travel_checklist.go
package main

import (
	"encoding/csv"
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
	"time"
)

const dataFile = "checklist.json"

type Item struct {
	ID       int    `json:"id"`
	Text     string `json:"text"`
	Category string `json:"category"`
	Packed   bool   `json:"packed"`
}

type Checklist struct {
	Name    string `json:"name"`
	Items   []Item `json:"items"`
	NextID  int    `json:"next_id"`
}

var templates = map[string]map[string][]string{
	"beach": {
		"Документы": {"Паспорт", "Билеты", "Страховка", "Виза"},
		"Одежда":    {"Купальник", "Шорты", "Футболка", "Панама", "Сандалии", "Пляжное полотенце"},
		"Гигиена":   {"Солнцезащитный крем", "Очки", "Шампунь", "Зубная щетка", "Гель для душа"},
		"Техника":   {"Телефон", "Зарядка", "Наушники", "Повербанк"},
		"Медицина":  {"Аптечка", "Средство от укусов", "Пластырь"},
	},
	"mountain": {
		"Документы":   {"Паспорт", "Страховка", "Карта", "Компас"},
		"Одежда":      {"Термобелье", "Флис", "Куртка", "Штаны", "Трекинговая обувь", "Носки", "Шапка", "Перчатки"},
		"Снаряжение":  {"Рюкзак", "Палатка", "Спальник", "Каремат", "Фонарик", "Трекинговые палки"},
		"Еда":         {"Горелка", "Топливо", "Посуда", "Еда в пакетах", "Вода", "Термос"},
		"Техника":     {"Телефон", "Зарядка", "Навигатор", "Радиостанция"},
	},
	"business": {
		"Документы": {"Паспорт", "Билеты", "Командировочное удостоверение", "Виза", "Страховка"},
		"Одежда":    {"Костюм", "Рубашки", "Галстук", "Туфли", "Носки", "Ремень"},
		"Техника":   {"Ноутбук", "Зарядка", "Презентация", "Флешка", "Телефон", "Адаптер"},
		"Гигиена":   {"Дезодорант", "Зубная щетка", "Паста", "Расческа", "Шампунь"},
		"Прочее":    {"Визитки", "Блокнот", "Ручка", "Зонт"},
	},
	"hiking": {
		"Документы":   {"Паспорт", "Карта", "Разрешение", "Страховка"},
		"Снаряжение":  {"Рюкзак", "Палатка", "Спальник", "Каремат", "Котелок", "Нож"},
		"Одежда":      {"Треккинговая обувь", "Носки", "Флис", "Куртка", "Штаны", "Дождевик", "Головной убор"},
		"Еда":         {"Еда в пакетах", "Сухой паек", "Вода", "Термос", "Горелка"},
		"Медицина":    {"Аптечка", "Эластичный бинт", "Пластырь", "Средство от насекомых"},
	},
	"cruise": {
		"Документы": {"Паспорт", "Виза", "Билеты", "Страховка", "Медицинская справка"},
		"Одежда":    {"Праздничный наряд", "Пляжная одежда", "Купальник", "Сандалии", "Туфли", "Пижама"},
		"Гигиена":   {"Солнцезащитный крем", "Шампунь", "Зубная щетка", "Дезодорант", "Москитная сетка"},
		"Техника":   {"Телефон", "Зарядка", "Фотоаппарат", "Наушники"},
		"Прочее":    {"Книга", "Ласты", "Маска для плавания", "Бинокль"},
	},
	"festival": {
		"Документы": {"Паспорт", "Билет на фестиваль", "Страховка"},
		"Одежда":    {"Яркая одежда", "Дождевик", "Удобная обувь", "Головной убор", "Солнцезащитные очки"},
		"Еда":       {"Вода", "Еда в пакетах", "Термос", "Пикник"},
		"Техника":   {"Телефон", "Зарядка", "Наушники", "Повербанк"},
		"Прочее":    {"Плед", "Стул", "Флаг", "Настольные игры"},
	},
}

func (c *Checklist) loadTemplate(name string) {
	tmpl, ok := templates[name]
	if !ok {
		return
	}
	c.Name = name
	c.Items = []Item{}
	c.NextID = 1
	for cat, items := range tmpl {
		for _, text := range items {
			c.Items = append(c.Items, Item{ID: c.NextID, Text: text, Category: cat, Packed: false})
			c.NextID++
		}
	}
}

func (c *Checklist) addItem(text, category string) {
	if category == "" {
		category = "Прочее"
	}
	item := Item{ID: c.NextID, Text: text, Category: category, Packed: false}
	c.Items = append(c.Items, item)
	c.NextID++
	fmt.Printf("\033[32mПункт добавлен (ID: %d)\033[0m\n", item.ID)
}

func (c *Checklist) removeItem(id int) {
	for i, item := range c.Items {
		if item.ID == id {
			c.Items = append(c.Items[:i], c.Items[i+1:]...)
			fmt.Printf("\033[33mПункт #%d удалён.\033[0m\n", id)
			return
		}
	}
	fmt.Printf("\033[31mПункт #%d не найден.\033[0m\n", id)
}

func (c *Checklist) packItem(id int) {
	for i := range c.Items {
		if c.Items[i].ID == id {
			c.Items[i].Packed = true
			fmt.Printf("\033[32mПункт #%d отмечен как собранный.\033[0m\n", id)
			return
		}
	}
	fmt.Printf("\033[31mПункт #%d не найден.\033[0m\n", id)
}

func (c *Checklist) unpackItem(id int) {
	for i := range c.Items {
		if c.Items[i].ID == id {
			c.Items[i].Packed = false
			fmt.Printf("\033[33mОтметка с пункта #%d снята.\033[0m\n", id)
			return
		}
	}
	fmt.Printf("\033[31mПункт #%d не найден.\033[0m\n", id)
}

func (c *Checklist) display() {
	if len(c.Items) == 0 {
		fmt.Println("\033[33mЧек-лист пуст.\033[0m")
		return
	}
	packed := 0
	for _, item := range c.Items {
		if item.Packed {
			packed++
		}
	}
	total := len(c.Items)
	pct := packed * 100 / total
	barLen := 20
	filled := pct * barLen / 100
	bar := strings.Repeat("█", filled) + strings.Repeat("░", barLen-filled)
	fmt.Printf("\033[36m📋 Чек-лист: %s\033[0m\n", c.Name)
	fmt.Printf("\033[32mПрогресс: %s %d%% (%d/%d)\033[0m\n\n", bar, pct, packed, total)

	categories := []string{}
	catMap := make(map[string]bool)
	for _, item := range c.Items {
		if !catMap[item.Category] {
			catMap[item.Category] = true
			categories = append(categories, item.Category)
		}
	}
	sort.Strings(categories)

	for _, cat := range categories {
		fmt.Printf("\033[33mКатегория: %s\033[0m\n", cat)
		for _, item := range c.Items {
			if item.Category == cat {
				status := "\033[32m✅\033[0m"
				if !item.Packed {
					status = "\033[31m❌\033[0m"
				}
				fmt.Printf("  %d. %s %s\n", item.ID, status, item.Text)
			}
		}
	}
}

func (c *Checklist) save(filename string) {
	data, _ := json.MarshalIndent(c, "", "  ")
	os.WriteFile(filename, data, 0644)
	fmt.Printf("\033[32mЧек-лист сохранён в %s\033[0m\n", filename)
}

func loadChecklist(filename string) *Checklist {
	data, err := os.ReadFile(filename)
	if err != nil {
		fmt.Printf("\033[31mОшибка загрузки: %v\033[0m\n", err)
		return nil
	}
	var c Checklist
	if err := json.Unmarshal(data, &c); err != nil {
		fmt.Printf("\033[31mОшибка парсинга: %v\033[0m\n", err)
		return nil
	}
	return &c
}

func (c *Checklist) exportCSV(filename string) {
	f, _ := os.Create(filename)
	defer f.Close()
	w := csv.NewWriter(f)
	defer w.Flush()
	w.Write([]string{"id", "category", "item", "packed"})
	for _, item := range c.Items {
		w.Write([]string{
			strconv.Itoa(item.ID),
			item.Category,
			item.Text,
			strconv.FormatBool(item.Packed),
		})
	}
	fmt.Printf("\033[32mЭкспортировано в %s (CSV)\033[0m\n", filename)
}

func (c *Checklist) exportTXT(filename string) {
	f, _ := os.Create(filename)
	defer f.Close()
	fmt.Fprintf(f, "Чек-лист: %s\n", c.Name)
	fmt.Fprintf(f, "Дата: %s\n\n", time.Now().Format("2006-01-02 15:04:05"))
	categories := []string{}
	catMap := make(map[string]bool)
	for _, item := range c.Items {
		if !catMap[item.Category] {
			catMap[item.Category] = true
			categories = append(categories, item.Category)
		}
	}
	sort.Strings(categories)
	for _, cat := range categories {
		fmt.Fprintf(f, "=== %s ===\n", cat)
		for _, item := range c.Items {
			if item.Category == cat {
				status := "[ ]"
				if item.Packed {
					status = "[x]"
				}
				fmt.Fprintf(f, "%s %s\n", status, item.Text)
			}
		}
		fmt.Fprintln(f)
	}
	fmt.Printf("\033[32mЭкспортировано в %s (TXT)\033[0m\n", filename)
}

func main() {
	var (
		template    string
		name        string
		add         string
		addCategory string
		pack        int
		unpack      int
		remove      int
		list        bool
		saveFile    string
		loadFile    string
		newFlag     bool
		exportCSV   string
		exportTXT   string
	)
	flag.StringVar(&template, "template", "", "Тип путешествия")
	flag.StringVar(&name, "name", "", "Название чек-листа")
	flag.StringVar(&add, "add", "", "Добавить пункт")
	flag.StringVar(&addCategory, "add-category", "", "Категория для добавляемого пункта")
	flag.IntVar(&pack, "pack", 0, "Отметить пункт как собранный")
	flag.IntVar(&unpack, "unpack", 0, "Снять отметку с пункта")
	flag.IntVar(&remove, "remove", 0, "Удалить пункт")
	flag.BoolVar(&list, "list", false, "Показать чек-лист")
	flag.StringVar(&saveFile, "save", "", "Сохранить чек-лист")
	flag.StringVar(&loadFile, "load", "", "Загрузить чек-лист")
	flag.BoolVar(&newFlag, "new", false, "Создать новый пустой чек-лист")
	flag.StringVar(&exportCSV, "export-csv", "", "Экспорт в CSV")
	flag.StringVar(&exportTXT, "export-txt", "", "Экспорт в TXT")
	flag.Parse()

	var c *Checklist
	if loadFile != "" {
		c = loadChecklist(loadFile)
		if c == nil {
			os.Exit(1)
		}
	} else if newFlag {
		c = &Checklist{Name: name, Items: []Item{}, NextID: 1}
	} else if template != "" {
		c = &Checklist{Name: name}
		c.loadTemplate(template)
	} else {
		// Попытка загрузить из файла по умолчанию
		if _, err := os.Stat(dataFile); err == nil {
			c = loadChecklist(dataFile)
		}
		if c == nil {
			c = &Checklist{Name: name, Items: []Item{}, NextID: 1}
		}
	}

	if add != "" {
		c.addItem(add, addCategory)
	}
	if pack != 0 {
		c.packItem(pack)
	}
	if unpack != 0 {
		c.unpackItem(unpack)
	}
	if remove != 0 {
		c.removeItem(remove)
	}
	if list {
		c.display()
	}
	if saveFile != "" {
		c.save(saveFile)
	}
	if exportCSV != "" {
		c.exportCSV(exportCSV)
	}
	if exportTXT != "" {
		c.exportTXT(exportTXT)
	}

	if loadFile == "" && !newFlag && template == "" && saveFile == "" && (add != "" || pack != 0 || unpack != 0 || remove != 0) {
		c.save(dataFile)
	}

	if flag.NArg() == 0 && len(os.Args) == 1 {
		fmt.Println("Используйте --help для справки.")
	}
}

File: travel_checklist.rs
// travel_checklist.rs
use clap::{App, Arg};
use colored::*;
use serde::{Deserialize, Serialize};
use serde_json;
use std::collections::{HashMap, HashSet};
use std::fs;
use std::io::Write;
use std::time::SystemTime;

const DATA_FILE: &str = "checklist.json";

#[derive(Serialize, Deserialize)]
struct Item {
    id: u32,
    text: String,
    category: String,
    packed: bool,
}

#[derive(Serialize, Deserialize)]
struct Checklist {
    name: String,
    items: Vec<Item>,
    next_id: u32,
}

impl Checklist {
    fn new(name: &str) -> Self {
        Checklist {
            name: name.to_string(),
            items: Vec::new(),
            next_id: 1,
        }
    }

    fn load_template(&mut self, template_name: &str) {
        let templates = get_templates();
        let tmpl = match templates.get(template_name) {
            Some(t) => t,
            None => return,
        };
        self.name = tmpl.name.clone();
        self.items = Vec::new();
        self.next_id = 1;
        for (category, items) in &tmpl.categories {
            for text in items {
                self.items.push(Item {
                    id: self.next_id,
                    text: text.clone(),
                    category: category.clone(),
                    packed: false,
                });
                self.next_id += 1;
            }
        }
    }

    fn add_item(&mut self, text: &str, category: &str) {
        let cat = if category.is_empty() { "Прочее" } else { category };
        let item = Item {
            id: self.next_id,
            text: text.to_string(),
            category: cat.to_string(),
            packed: false,
        };
        self.items.push(item);
        self.next_id += 1;
        println!("{}", format!("Пункт добавлен (ID: {})", self.next_id - 1).green());
    }

    fn remove_item(&mut self, id: u32) {
        if let Some(pos) = self.items.iter().position(|i| i.id == id) {
            self.items.remove(pos);
            println!("{}", format!("Пункт #{} удалён.", id).yellow());
        } else {
            println!("{}", format!("Пункт #{} не найден.", id).red());
        }
    }

    fn pack_item(&mut self, id: u32) {
        if let Some(item) = self.items.iter_mut().find(|i| i.id == id) {
            item.packed = true;
            println!("{}", format!("Пункт #{} отмечен как собранный.", id).green());
        } else {
            println!("{}", format!("Пункт #{} не найден.", id).red());
        }
    }

    fn unpack_item(&mut self, id: u32) {
        if let Some(item) = self.items.iter_mut().find(|i| i.id == id) {
            item.packed = false;
            println!("{}", format!("Отметка с пункта #{} снята.", id).yellow());
        } else {
            println!("{}", format!("Пункт #{} не найден.", id).red());
        }
    }

    fn display(&self) {
        if self.items.is_empty() {
            println!("{}", "Чек-лист пуст.".yellow());
            return;
        }
        let total = self.items.len();
        let packed = self.items.iter().filter(|i| i.packed).count();
        let pct = if total > 0 { packed * 100 / total } else { 0 };
        let bar_len = 20;
        let filled = pct * bar_len / 100;
        let bar = "█".repeat(filled) + "░".repeat(bar_len - filled);
        println!("{}", format!("📋 Чек-лист: {}", self.name).cyan());
        println!("{}", format!("Прогресс: {} {}% ({}/{})", bar, pct, packed, total).green());
        println!();

        let mut categories: Vec<String> = self.items.iter().map(|i| i.category.clone()).collect();
        categories.sort();
        categories.dedup();

        for cat in categories {
            println!("{}", format!("Категория: {}", cat).yellow());
            for item in self.items.iter().filter(|i| i.category == cat) {
                let status = if item.packed { "✅".green() } else { "❌".red() };
                println!("  {}. {} {}", item.id, status, item.text);
            }
        }
    }

    fn save(&self, filename: &str) {
        let json = serde_json::to_string_pretty(&self).unwrap();
        fs::write(filename, json).unwrap();
        println!("{}", format!("Чек-лист сохранён в {}", filename).green());
    }

    fn load(filename: &str) -> Option<Self> {
        if let Ok(data) = fs::read_to_string(filename) {
            if let Ok(checklist) = serde_json::from_str(&data) {
                return Some(checklist);
            }
        }
        println!("{}", "Ошибка загрузки.".red());
        None
    }

    fn export_csv(&self, filename: &str) {
        let mut wtr = csv::Writer::from_path(filename).unwrap();
        wtr.write_record(&["id", "category", "item", "packed"]).unwrap();
        for item in &self.items {
            wtr.write_record(&[
                &item.id.to_string(),
                &item.category,
                &item.text,
                &item.packed.to_string(),
            ]).unwrap();
        }
        wtr.flush().unwrap();
        println!("{}", format!("Экспортировано в {} (CSV)", filename).green());
    }

    fn export_txt(&self, filename: &str) {
        let mut content = format!("Чек-лист: {}\n", self.name);
        content.push_str(&format!("Дата: {:?}\n\n", SystemTime::now()));
        let mut categories: Vec<String> = self.items.iter().map(|i| i.category.clone()).collect();
        categories.sort();
        categories.dedup();
        for cat in categories {
            content.push_str(&format!("=== {} ===\n", cat));
            for item in self.items.iter().filter(|i| i.category == cat) {
                let status = if item.packed { "[x]" } else { "[ ]" };
                content.push_str(&format!("{} {}\n", status, item.text));
            }
            content.push('\n');
        }
        fs::write(filename, content).unwrap();
        println!("{}", format!("Экспортировано в {} (TXT)", filename).green());
    }
}

struct Template {
    name: String,
    categories: HashMap<String, Vec<String>>,
}

fn get_templates() -> HashMap<String, Template> {
    let mut templates = HashMap::new();
    
    let mut beach_cats = HashMap::new();
    beach_cats.insert("Документы".to_string(), vec!["Паспорт".to_string(), "Билеты".to_string(), "Страховка".to_string(), "Виза".to_string()]);
    beach_cats.insert("Одежда".to_string(), vec!["Купальник".to_string(), "Шорты".to_string(), "Футболка".to_string(), "Панама".to_string(), "Сандалии".to_string(), "Пляжное полотенце".to_string()]);
    beach_cats.insert("Гигиена".to_string(), vec!["Солнцезащитный крем".to_string(), "Очки".to_string(), "Шампунь".to_string(), "Зубная щетка".to_string(), "Гель для душа".to_string()]);
    beach_cats.insert("Техника".to_string(), vec!["Телефон".to_string(), "Зарядка".to_string(), "Наушники".to_string(), "Повербанк".to_string()]);
    beach_cats.insert("Медицина".to_string(), vec!["Аптечка".to_string(), "Средство от укусов".to_string(), "Пластырь".to_string()]);
    templates.insert("beach".to_string(), Template { name: "Пляжный отдых".to_string(), categories: beach_cats });

    let mut mountain_cats = HashMap::new();
    mountain_cats.insert("Документы".to_string(), vec!["Паспорт".to_string(), "Страховка".to_string(), "Карта".to_string(), "Компас".to_string()]);
    mountain_cats.insert("Одежда".to_string(), vec!["Термобелье".to_string(), "Флис".to_string(), "Куртка".to_string(), "Штаны".to_string(), "Трекинговая обувь".to_string(), "Носки".to_string(), "Шапка".to_string(), "Перчатки".to_string()]);
    mountain_cats.insert("Снаряжение".to_string(), vec!["Рюкзак".to_string(), "Палатка".to_string(), "Спальник".to_string(), "Каремат".to_string(), "Фонарик".to_string(), "Трекинговые палки".to_string()]);
    mountain_cats.insert("Еда".to_string(), vec!["Горелка".to_string(), "Топливо".to_string(), "Посуда".to_string(), "Еда в пакетах".to_string(), "Вода".to_string(), "Термос".to_string()]);
    mountain_cats.insert("Техника".to_string(), vec!["Телефон".to_string(), "Зарядка".to_string(), "Навигатор".to_string(), "Радиостанция".to_string()]);
    templates.insert("mountain".to_string(), Template { name: "Горный поход".to_string(), categories: mountain_cats });

    // ... (остальные шаблоны аналогично, сокращённо для краткости)
    let mut business_cats = HashMap::new();
    business_cats.insert("Документы".to_string(), vec!["Паспорт".to_string(), "Билеты".to_string(), "Командировочное удостоверение".to_string(), "Виза".to_string(), "Страховка".to_string()]);
    business_cats.insert("Одежда".to_string(), vec!["Костюм".to_string(), "Рубашки".to_string(), "Галстук".to_string(), "Туфли".to_string(), "Носки".to_string(), "Ремень".to_string()]);
    business_cats.insert("Техника".to_string(), vec!["Ноутбук".to_string(), "Зарядка".to_string(), "Презентация".to_string(), "Флешка".to_string(), "Телефон".to_string(), "Адаптер".to_string()]);
    business_cats.insert("Гигиена".to_string(), vec!["Дезодорант".to_string(), "Зубная щетка".to_string(), "Паста".to_string(), "Расческа".to_string(), "Шампунь".to_string()]);
    business_cats.insert("Прочее".to_string(), vec!["Визитки".to_string(), "Блокнот".to_string(), "Ручка".to_string(), "Зонт".to_string()]);
    templates.insert("business".to_string(), Template { name: "Командировка".to_string(), categories: business_cats });

    templates
}

fn main() {
    let matches = App::new("Travel Checklist")
        .arg(Arg::with_name("template").long("template").takes_value(true).help("Тип путешествия"))
        .arg(Arg::with_name("name").long("name").takes_value(true).help("Название чек-листа"))
        .arg(Arg::with_name("add").long("add").takes_value(true).help("Добавить пункт"))
        .arg(Arg::with_name("add-category").long("add-category").takes_value(true).help("Категория для добавляемого пункта"))
        .arg(Arg::with_name("pack").long("pack").takes_value(true).help("Отметить пункт как собранный"))
        .arg(Arg::with_name("unpack").long("unpack").takes_value(true).help("Снять отметку с пункта"))
        .arg(Arg::with_name("remove").long("remove").takes_value(true).help("Удалить пункт"))
        .arg(Arg::with_name("list").long("list").help("Показать чек-лист"))
        .arg(Arg::with_name("save").long("save").takes_value(true).help("Сохранить чек-лист"))
        .arg(Arg::with_name("load").long("load").takes_value(true).help("Загрузить чек-лист"))
        .arg(Arg::with_name("new").long("new").help("Создать новый пустой чек-лист"))
        .arg(Arg::with_name("export-csv").long("export-csv").takes_value(true).help("Экспорт в CSV"))
        .arg(Arg::with_name("export-txt").long("export-txt").takes_value(true).help("Экспорт в TXT"))
        .get_matches();

    let mut checklist = if let Some(file) = matches.value_of("load") {
        if let Some(c) = Checklist::load(file) { c } else { std::process::exit(1); }
    } else if matches.is_present("new") {
        Checklist::new(matches.value_of("name").unwrap_or("Новый список"))
    } else if let Some(tmpl) = matches.value_of("template") {
        let mut c = Checklist::new(matches.value_of("name").unwrap_or("Путешествие"));
        c.load_template(tmpl);
        c
    } else {
        if let Some(c) = Checklist::load(DATA_FILE) { c } else { Checklist::new("Мой список") }
    };

    if let Some(text) = matches.value_of("add") {
        let category = matches.value_of("add-category").unwrap_or("Прочее");
        checklist.add_item(text, category);
    }
    if let Some(id_str) = matches.value_of("pack") {
        checklist.pack_item(id_str.parse().unwrap());
    }
    if let Some(id_str) = matches.value_of("unpack") {
        checklist.unpack_item(id_str.parse().unwrap());
    }
    if let Some(id_str) = matches.value_of("remove") {
        checklist.remove_item(id_str.parse().unwrap());
    }
    if matches.is_present("list") {
        checklist.display();
    }
    if let Some(file) = matches.value_of("save") {
        checklist.save(file);
    }
    if let Some(file) = matches.value_of("export-csv") {
        checklist.export_csv(file);
    }
    if let Some(file) = matches.value_of("export-txt") {
        checklist.export_txt(file);
    }

    // Автосохранение
    if !matches.is_present("load") && !matches.is_present("new") && matches.value_of("template").is_none() &&
       !matches.is_present("save") && (matches.is_present("add") || matches.is_present("pack") ||
       matches.is_present("unpack") || matches.is_present("remove")) {
        checklist.save(DATA_FILE);
    }

    if matches.args_present() == 0 {
        println!("Используйте --help для справки.");
    }
}
